lib/spack/spack/solver/asp.py

```diff
@@ -1938,11 +1953,28 @@ def setup(self, driver, specs, reuse=None):
         # rules to generate an ASP program.
         self.gen = driver
 
+        # Calculate develop specs
+        # they will be used in addition to command line specs
+        # in determining known versions/targets/os
+        env = ev.active_environment()
+        dev_specs = (
+            tuple(
+                spack.spec.Spec(info["spec"]).constrained(
+                    "dev_path=%s"
+                    % spack.util.path.canonicalize_path(info["path"], default_wd=env.path)
+                )
+                for name, info in env.dev_specs.items()
+            )
+            if env
+            else tuple()
+        )
+        specs = tuple(specs)  # ensure compatible types to add
+
         # get possible compilers
         self.possible_compilers = self.generate_possible_compilers(specs)
 
         # traverse all specs and packages to build dict of possible versions
-        self.build_version_dict(possible, specs)
+        self.build_version_dict(possible, specs, dev_specs)
```

Comment:
to make this more clear I would factor out the bottom half of `build_version_dict` as I mentioned above, but call it here:

```python
    self.build_version_dict(possible)
    self.add_concrete_versions_from_specs(specs, version_provenance.spec)
    self.add_concrete_versions_from_specs(dev_specs, version_provenance.dev_spec)
```

That function is doing too much stuff -- the two halves are really two functions and conflating the parameters here makes the code hard to follow.